### PR TITLE
Ruby group capstone - Catalog of my things🟢

### PR DESCRIPTION
**Pull request by SabaAhmad404**
we are a team of three people

- @teke85 
- @SabaAhmad404 
- @haliljon 

- ### we have divided project into three members

👇 👇 👇 
## [Project Documentation video link](https://drive.google.com/file/d/11DdE-O2xQMBYlS8F2AV0QOcKWEtU8r3E/view?usp=sharing)
👆 👆 👆 

**1.  Haliljon Juraboev#1**

- Create a Book class in a separate .rb file.

- Create a Label class with an association to the Item class (in a separate .rb file).

- All Book class properties visible in the diagram should be defined and set up in the constructor method.
- All Label class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
 

>  add_item method in the Label class

> 1. should take an instance of the Item class as an input
> 2. should add the input item to the collection of items
> 3. should add self as a property of the item object (by using the correct setter from the item object)

> can_be_archived?() in the Book class

> 1. should override the method from the parent class
> 2. should return true if parent's method returns true OR if cover_state equals to "bad"
> 3. otherwise, it should return false

- Add unit tests for all implemented methods.

- The following options should be available:

> 1. List all books
> 2. List all labels (e.g. 'Gift', 'New')
> 3. Add a book

- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
books table (add all properties and associations from the parent Item class as table columns)
labels table

**2. Saba Ahmad #2**

- Create the MusicAlbum class in a separate .rb file.
- Create a Genre class with an association to the Item class (in a separate .rb file).
- All MusicAlbum class properties visible in the diagram should be defined and set up in the constructor method.
- All Genre class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
> add_item method in the Genre class

> 1. should take an instance of the Item class as an input
> 2. should add the input item to the collection of items
> 3. should add self as a property of the item object (by using the correct setter from the item object)

> can_be_archived?() in the MusicAlbum class

> 1. should override the method from the parent class
> 2. should return true if parent's method returns true AND if on_spotify equals true
> 3. otherwise, it should return false
- Add unit tests for all implemented methods.
- The following options should be available:

> 1. List all music albums
> 2. List all genres (e.g 'Comedy', 'Thriller')
> 3. Add a music album
- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
music_albums table (add all properties and associations from the parent Item class as table columns)
genres table

**3. Isiteketo Mutau #3** 

- Create a Game class in a separate .rb file.
- Create an Author class with an association to the Item class (in a separate .rb file).
- All Game class properties visible in the diagram should be defined and set up in the constructor method.
- All Author class properties visible in the diagram should be defined and set up in the constructor method.
- Implement methods:
> add_item method in the Author class

> 1. should take an instance of the Item class as an input
> 2. should add the input item to the collection of items
> 3. should add self as a property of the item object (by using the correct setter from the item object)

> can_be_archived?() in the Game class

> 1. should override the method from the parent class
> 2. should return true if parent's method returns true AND if last_played_at is older than 2 years
> 3. otherwise, it should return false

- Add unit tests for all implemented methods.
- The following options should be available:

> 1. List of games
> 2. List all authors (e.g. 'Stephen King')
> 3. Add a game

- All data should be preserved by saving collections in .json files.
- Create a schema.sql file with tables that will be analogical to the structure of the classes that you created:
games table (add all properties and associations from the parent Item class as table columns)
authors table

### CODE REVIEW

File: spec/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      expect(book2.can_be_archived?).to be true
+    end
+
+    it 'should return false if book is not older than 10 years and cover state is good' do
+      expect(book3.can_be_archived?).to be false
+    end
+
+    it 'should return true if book is not older than 10 years and cover state is bad' do
+      expect(book4.can_be_archived?).to be true
+    end
```
Comment (danmainah) [inline spec/book_spec.rb:42]: - [ ] @team great work on your test however you forgot to add a test for `to_json(*args)` and `initialize` methods, kindly add tests for these methods

File: spec/label_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    it 'should add item to items array' do
+      label.add_item(item)
+      expect(label.items).to include(item)
+    end
+
+    it 'should add book to items array' do
+      label.add_item(book)
+      expect(label.items).to include(book)
+    end
+  end
```
Comment (danmainah) [inline spec/label_spec.rb:36]: - [ ]  @team great work on your test however you forgot to add a test for to_json(*args) and initialize methods, kindly add tests for these methods

File: spec/game_spec.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+describe Game do
+  context '#can_be_archive when called' do
+    it 'should return false if @last_played_at > 2' do
+      game = Game.new('2010-03-10', 'yes', '2022-03-21')
+      result = game.can_be_archived?
+
+      expect(result).to eq(false)
+    end
+  end
```
Comment (danmainah) [inline spec/game_spec.rb:23]: - [ ]  @team great work on your test however you forgot to add a test for initialize method, kindly add tests for this method.

File: spec/music_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  context 'Test MusicAlbum with methods' do
+    it 'Expecting new instance' do
+      expect(@music).to be_instance_of(MusicAlbum)
+    end
+  end
+
+  it 'Expecting can_be_archive? method' do
+    expect(@music.can_be_archived?).to be(true)
+  end
```
Comment (danmainah) [inline spec/music_spec.rb]: - [ ] @team great work on your tests however you forgot to add tests for the item and music classes, kindly ensure to test all public methods in these classes.

File: classes/book/booklabelmodule.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  def store_book(book)
+    books = fetch_books
+    books << book
+    File.write('./classes/book/books.json', JSON.pretty_generate(books), mode: 'w')
+  end
+
+  def store_label(label)
+    labels = fetch_labels
+    labels << label
+    File.write('./classes/book/labels.json', JSON.pretty_generate(labels), mode: 'w')
```
Comment (danmainah) [inline classes/book/booklabelmodule.rb:61]: - [ ] @team am getting the error below when adding a new book, although the book gets added,I can't add a label , kindly fix the error
image

File: README.md [hunk 6]
```diff
@@ -1 +1,154 @@
...
+Clone this repository to your desired folder:
+
+- Open the command prompt in the created directory.
+
+- On the terminal run this command git clone https://github.com/SabaAhmad404/Ruby-Capstone.git
+
+- Go to the repository folder using command prompt cd OOP-School-Library
+
+- Install the dev dependencies for linters run npm install.
+
```
Comment (danmainah) [inline README.md:93]: - [ ] @team kindly add instructions on how to run your app and run tests in it.

File: schema/schema.sql [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    label_id INTEGER,
+    author_id INTEGER,
+    source_id INTEGER,
+    archived BOOLEAN,
+    multiplayer VARCHAR(5)
+    last_played_at DATE,
+    FOREIGN KEY (genre_id) REFERENCES Genre (id),
+    FOREIGN KEY (author_id) REFERENCES Authors (id),
+    FOREIGN KEY (label_id) REFERENCES Label (id),
+);
```
Comment (danmainah) [inline schema/schema.sql]: - [x] @team great work on your schema however you forgot to add indexes for foreign keys, kindly add indexes for all your foreign keys

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @SabaAhmad404, @haliljon and @teke85 , 

Good job so far!
You still need to work on some issues to go to the next project, but you are almost there!

## Highlights
✔️ No linter errors
✔️ Gitflow was correctly used
✔️  Good design


# Required Changes ♻️

- [x] @team great work on your project, however, you forgot to add a project presentation, kindly add a presentation and ensure it follows all the requirements below
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/book_spec.rb (hunk 1 above)
Comment (teke85) [reply spec/book_spec.rb:42]: Hi @danmainah thank you for the comment. We have added the ```to_json(*args)``` and ```initialize``` methods to our tests

File: spec/label_spec.rb (hunk 2 above)
Comment (teke85) [reply spec/label_spec.rb:36]: Hi @danmainah thank you for the comment. We have added the ```to_json(*args)``` and ```initialize``` methods to our tests

File: spec/game_spec.rb (hunk 3 above)
Comment (teke85) [reply spec/game_spec.rb:23]: Hi @danmainah thank you for the comment. We have added the ```initialize``` method to our tests

File: spec/music_spec.rb (hunk 4 above)
Comment (teke85) [reply spec/music_spec.rb]: Hi @danmainah thank you for the comment. We have added the ```item``` and ```music``` classes.

File: classes/book/booklabelmodule.rb (hunk 5 above)
Comment (teke85) [reply classes/book/booklabelmodule.rb:61]: Hi @danmainah thank you for the comment. We have fixed the error the app is working now.

File: README.md (hunk 6 above)
Comment (teke85) [reply README.md:93]: Hi @danmainah thank you for the comment. We have updated our readme file

**Comment by haliljon** [comment]
> Hi @SabaAhmad404, @haliljon and @teke85 ,
> 
> Good job so far! You still need to work on some issues to go to the next project, but you are almost there!
> 
> ## Highlights
> ✔️ No linter errors ✔️ Gitflow was correctly used ✔️ Good design
> 
> # Required Changes ♻️
> * [x]  @team great work on your project, however, you forgot to add a project presentation, kindly add a presentation and ensure it follows all the requirements below
>   image
> 
> _Check the comments under the review._
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

Hi @danmainah 

We just added the "Project Documentaion video" to the PR, thank you 🙏

**Review by TadesseDev — APPROVED** [review]
### Hi Team

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
giphy (1)



## Highlights
 - [x] Good job making the changes requested by the previous code reviewer. ✅
## Optional suggestions
- [Optional] You have done a great job team but please notice for the `Video presentation` we need to meet certain requirements like `maximum length of the video is 5 ❌`, `video contains a clear and objective description of the project in professional English ✅`,  `contains a demo of the project ✅` ,  `contains a highlight about a piece of the code you finds interesting or you are proud of making ✅`. and your documentation is missing the time limit. As a professional developer, we need to adapt to document our work professionally. 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**